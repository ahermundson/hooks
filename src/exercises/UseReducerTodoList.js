import React from "react";
import uuid from "uuid";
import TodoListForm from "./UseStateTodoList/TodoListForm";
import TodoListItem from "./UseStateTodoList/TodoListItem";

const UseReducerTodoList = () => {
  // TODO: Replace useState with useReducer. HINT: React.useReducer expects a reducer function and an initial state
  const [todoInputValue, updateTodoInputValue] = React.useState("");
  const [todos, updateTodos] = React.useState([]);

  // Replace function calls with dispatches
  function handleTodoInputChange(e) {
    updateTodoInputValue(e.target.value);
  }

  function onSubmit() {
    updateTodos([...todos, { todo: todoInputValue, id: uuid() }]);
    updateTodoInputValue("");
  }

  return (
    <React.Fragment>
      <TodoListForm
        todoInputValue={todoInputValue}
        handleSubmit={onSubmit}
        handleTodoInputChange={handleTodoInputChange}
        header="Use Reducer"
      />
      <ul data-testid="todo-list">
        {todos.map(({ todo, id }) => (
          <TodoListItem key={id} todo={todo} />
        ))}
      </ul>
    </React.Fragment>
  );
};

export default UseReducerTodoList;
