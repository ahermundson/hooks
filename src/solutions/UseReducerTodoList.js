import React from "react";
import uuid from "uuid";
import TodoListForm from "../exercises/UseStateTodoList/TodoListForm";
import TodoListItem from "../exercises/UseStateTodoList/TodoListItem";

const UseReducerTodoList = () => {
  // TODO: Replace useState with useReducer. HINT: React.useReducer expects a reducer function and an initial state
  const initialState = {
    todoInputValue: "",
    todos: []
  };

  function reducer(state, action) {
    switch (action.type) {
      case "HANDLE_TODO_INPUT_CHANGE":
        return {
          ...state,
          todoInputValue: action.payload
        };
      case "ADD_TODO":
        return {
          todoInputValue: "",
          todos: [...state.todos, { todo: state.todoInputValue, id: uuid() }]
        };
      default:
        return state;
    }
  }

  const [state, dispatch] = React.useReducer(reducer, initialState);

  // Replace function calls with dispatches
  function handleTodoInputChange(e) {
    dispatch({ type: "HANDLE_TODO_INPUT_CHANGE", payload: e.target.value });
  }

  function onSubmit() {
    dispatch({ type: "ADD_TODO" });
  }

  return (
    <React.Fragment>
      <TodoListForm
        todoInputValue={state.todoInputValue}
        handleSubmit={onSubmit}
        handleTodoInputChange={handleTodoInputChange}
        header="Use Reducer"
      />
      <ul data-testid="todo-list">
        {state.todos.map(({ todo, id }) => (
          <TodoListItem key={id} todo={todo} />
        ))}
      </ul>
    </React.Fragment>
  );
};

export default UseReducerTodoList;
