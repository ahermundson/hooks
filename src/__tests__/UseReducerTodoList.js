import React from "react";
import { render, fireEvent } from "react-testing-library";
import UseReducerTodoList from "../exercises/UseReducerTodoList";
// import UseReducerTodoList from "../solutions/UseReducerTodoList";

test("Can add a todo item - UseReducerTodoList", () => {
  const { container, getByPlaceholderText, getByText } = render(
    <UseReducerTodoList />
  );
  const todoInput = getByPlaceholderText("Add Todo");
  const submitButton = getByText("Submit");

  fireEvent.change(todoInput, { target: { value: "todo" } });
  expect(todoInput.value).toBe("todo");
  fireEvent.click(submitButton);
  expect(todoInput.value).toBe("");

  expect(container).toMatchSnapshot();
});
