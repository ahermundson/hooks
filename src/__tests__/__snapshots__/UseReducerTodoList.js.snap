// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can add a todo item - UseReducerTodoList 1`] = `
<div>
  <form
    class="login-form"
  >
    <h1
      style="text-align: center;"
    >
      Use Reducer
    </h1>
    <div
      class="input-group"
    >
      <label
        for="username-input"
      >
        Todo
      </label>
      <input
        id="todo-input"
        placeholder="Add Todo"
        type="text"
        value=""
      />
    </div>
    <button>
      Submit
    </button>
  </form>
  <ul
    data-testid="todo-list"
  >
    <li
      data-testid="list-item"
    >
      todo
    </li>
  </ul>
</div>
`;
